### ISSUE TITLE
Rethink if I really need a rectangular grid in room generation

### ISSUE THREAD

**Issue by Mikolaj** [LambdaHack/LambdaHack#187]
of perhaps a BSP partition and a graph of adjacency would be enough, as in the following post. The rethinking would ideally materialize as a re-implementation. This is hard, but at least self-contained, because the rest of the engine doesn't know how the level map was obtained, except for knowing which tiles are room entrances, for debug and extra hints to the player (and possibly AI in the future). 

https://www.reddit.com/r/gamedev/comments/dx95df/cave_generation_using_bsp_and_cellular_automaton/

**Comment by Mikolaj** [comment]
Or, a related question: do I really want to decide the size or rooms and then create a room, or rather create a room and find a place for it on the map? Given that I have small maps, probably I need the method that packs rooms most densely. The flexibility of resizing rooms is probably best for that, so partitioning the level first (but not as a grid) seems to be best.

**Comment by MichaelMackus** [comment]
I'm kinda running into this issue since it looks like all the current maps are created via the grid method. I was originally planning on using cellular automata but I'm not sure that's supported here. Perhaps we should abstract this one level higher, and offer a way to generate the base level via a function which returns the tiles/places? We could still keep the current "CaveKind" for generator-based grids in my mind, or even just add a new function to that data type for this generator idea.

Just spit-balling, feel free to update if there's a better/easier method.

**Comment by Mikolaj** [comment]
It depends on what exactly you need: a different map partitioning or different building of places inside map partition cells. Or, the hardest, if you want the former to be partially determined by the latter and still keep all the constraints of aligned stairs, distance between stairs, connectivity of stairs, of level border wall, proportion of walkable to non-walkable tiles, etc. Or, the most invasive change: if you'd like to drop the constraints sometimes and have dungeon sections with different sets of constraints.

We could relatively easily generate a map partition (e.g., the current method resulting in a grid) and then use cellular automata to populate it, operating on the level of partition cells (so the output of an automaton would assignment of place kinds, not tile kinds) and/or running inside some or all of the cells.

We could also relatively easily change the method for generating the partition (this issue it about that), while keeping all the constraints.

Yet another relatively easy option is to get the stairs from level above, generate partition globally, use the partition to place the stairs down and then forget the partition and fill the remaining space with cellular automata. The major difficulty is to have the automata respect the global constraints and. e.g., connect all the stairs. But I'm sure there are known methods for that, e.g., starting one automaton per stairs, stopping automatons when close enough and then using wave function collapse to meld the paths they generated into connected graph. Or just fixing up brutally any violated constrain or re-generating until constraints met randomly.

**Comment by Mikolaj** [comment]
And regarding your original questions, it depends on whether you'd like to use LambdaHack as a library or modify it. Given your ambitious ideas, I'd say modifying it would be much easier (otherwise, you'd need to copy-paste most of `Server/StartM.hs` and `Server/LoopM.hs`, until we have a system in place that make swapping their components easier.

And if you are modifying, the easiest would be to add an extra filed to `CaveKind` which says if it's `FillGridBased` or `FillCellurarAutomata` and any other extra fields as needed, depending on what exactly is your plan. Most of the other fields would or could still apply. E.g., the grid parameters would be translatable to automate parametrs controlling how often and how large and how varied open spaces are generated (whether eventually filled with scaled prefabs or not, which could be controlled by another option).

**Comment by Mikolaj** [comment]
To put it in yet another way: you can define the cave as one big place and add a cellular method of to the (1-element) list of methods of place generation or you can define a cave that has no rooms and extend the (2-element with many extra parameters) list of methods of filling space between places. Optionally generate only stair places. e.g., the tiny ones. These are harder than starting from scratch, but you get a lot of stuff for free later on by integrating with what's already there.

**Comment by MichaelMackus** [comment]
I like your idea here of having different fill algorithms for grid pieces. It could even be cool if this could be abstracted further outside the engine, so the implementation can define its own algorithm completely. It would be good not to be constrained to grid-based algorithms, but this would be a great step.

**Comment by Mikolaj** [comment]
Yes, we can think of making dungeon generation code an argument of `loopSer`, passed to `gameReset`. Or go full-hog and specify all major building blocks of the engine in `RuleKind` as functions and whenever they are needed, they would be looked up in content. For large scale functions that are rarely called (just as dungeon generation, called just once per game), this is free. Only for small functions that are called in a loop (e.g., a function that would either allow 8-way or 4-way movement), this is costly, because calling unknown functions is costly. Worth opening an issue.

We can also divide the current dungeon generation functions further to make it easier to reuse them in the custom functions for a specific game. Also worth an ticket.
